templates/nextflow-s3-template.yaml

```diff
@@ -0,0 +1,148 @@
+
+# Mappings
+Mappings:
+  Accounts:
+    "035458030717":
+      AdminRole: "arn:aws:iam::035458030717:role/aws-reserved/sso.amazonaws.com/AWSReservedSSO_Administrator_580e9f32ac55c4e7"
+    "728882028485":
+      AdminRole: "arn:aws:iam::728882028485:role/aws-reserved/sso.amazonaws.com/AWSReservedSSO_Administrator_3a9d61d0884be260"
+
+# Resources
+Resources:
+  Bucket:
+    Type: AWS::S3::Bucket
+    Properties:
+      BucketName: !If [HasBucketName, !Ref BucketName, !Ref 'AWS::NoValue']
+      AccessControl: Private
+      BucketEncryption:
+        ServerSideEncryptionConfiguration:
+          - ServerSideEncryptionByDefault:
+              SSEAlgorithm: AES256
+
+  # Policy
+  BucketPolicy:
+    Type: AWS::S3::BucketPolicy
+    Properties:
+      Bucket: !Ref Bucket  
+      PolicyDocument:
+        Version: 2012-10-17
+        Statement:
+
+          - Sid: AllowDeleteAllowModify
+            # denies all except modifying, deleting and emptying bucket
+            Effect: Deny
+            Principal: '*'
+            NotAction:
+              - "s3:Delete*"
+              - "s3:GetBucket*"
+              - "s3:ListBucket*"
+              - "s3:PutBucket*"
+            Resource:
+              - !Sub arn:aws:s3:::${Bucket}
+              - !Sub arn:aws:s3:::${Bucket}/*
+
+          - Sid: AdminFullAccess
+            # gives admin full access to bucket
+            Effect: Allow
+            Principal:
+              AWS:
+                - !Sub "arn:aws:iam::${AWS::AccountId}:root"
+                - Fn::ImportValue: !Sub "${AWS::Region}-workflows-nextflow-ci-service-account-ServiceRoleArn"
+                - !FindInMap [Accounts, !Ref AWS::AccountId, AdminRole]
+            Action:
+              - "s3:*"
+            Resource:
+              - !Sub arn:aws:s3:::${Bucket}
+              - !Sub arn:aws:s3:::${Bucket}/*
+
+          - Sid: ReadWriteAccess
+            # gives grantees read/write access to the bucket
+            Effect: Allow
+            Principal:
+              AWS: !Split
+                - ","
+                - !Join
+                    - ","
+                    - - Fn::ImportValue: !Sub "${AWS::Region}-nextflow-forge-service-user-ServiceUserArn"
+                      - !Join
+                          - ","
+                          - !Ref ReadWriteAccess
+            Action:
+              - "s3:*Object*"
+              - "s3:*MultipartUpload*"
+              - "s3:ListBucket*"
+              - "s3:GetBucketLocation"
+            Resource:
+              - !Sub arn:aws:s3:::${Bucket}/*
+              - !Sub arn:aws:s3:::${Bucket}            
+
+          - Sid: ReadOnlyAccess
+            # gives grantees read (only) access to the bucket
+            Effect: Allow
+            Principal:
+              AWS: !Split
+                - ","
+                - !Join
+                    - ","
+                    - - Fn::ImportValue: !Sub "${AWS::Region}-nextflow-forge-service-user-ServiceUserArn"
```

Comment:
Why are we including the Forge user on both the `ReadWriteAccess` and `ReadOnlyAccess` statements? Isn't it the latter redundant? Or is this being done just for simplicity in case no values are given to the `ReadOnlyAccess` parameter?

Reply:
Forge user has been removed from ReadOnly! The !Join function doesn't work when the `ReadOnlyAccessArns` is left empty (in which case, should the Forge user be set as a default value for the ReadOnlyAccessArns parameter?)